### ISSUE TITLE
[TFLite] Built-in way to run the reference kernels with the Python API

### ISSUE THREAD

**Issue by Tessil** [tensorflow/tensorflow#44043]
Hello,

When accelerating inference (via optimized software or hardware implementations) we often want to compare the result accuracy with the reference result defined by TFLite reference kernels (`kReference`) instead of the default kernels (which often are `kGenericOptimized`).

With the C++ API it is easy to pass the `tflite::ops::builtin::BuiltinRefOpResolver` instead of the `tflite::ops::builtin::BuiltinOpResolver` to the `tflite::InterpreterBuilder` constructor. Unfortunately with the Python API we need to recompile TensorFlow to expose the experimental `tf.lite.InterpreterWithCustomOps` API and manually export the symbol of a C function which adds the `tflite::ops::builtin::BuiltinRefOpResolver` to the current `tflite::MutableOpResolver`.

As manually compiling TensorFlow instead of using the public binaries is quite cumbersome for some of our users, we would like to see if it would be possible to add a built-in way to run a model with the reference kernels using the publicly distributed binaries.

Thanks
Thibaut

**Comment by jdduke** [comment]
This seems like a reasonable request. We probably wouldn't expose this for other language bindings, like Java/Swift, but for Python it probably makes sense as that's "close" to the conversion code and could be useful for development, as you noted. 

Are you able to tell what the relative binary size difference would be if we linked in both the reference and optimized OpResolver into the binary?

**Comment by Tessil** [comment]
Comparing the full Linux TensorFlow 2.3 CPU I built with and without the reference kernels compiled into it, I have the following sizes.
* Base: 127 157 306 bytes
* With reference kernels: 127 276 271 bytes

So a difference of around 116 KiB (~0.09% relative increase, would be less for the GPU build).

**Comment by jdduke** [comment]
That doesn't seem too bad. Have you thought about what the API would look like when creating/using the Interpreter?

**Comment by Tessil** [comment]
I see a few way:
* Either adding an extra boolean or enum parameter to select if we want to use the optimized or reference kernels in `tf.lite.Interpreter` (default to optimized). Though it may make the interface more complex than needed for a feature that will not be used by many users.
* Currently there is the experimental [`tf.lite.InterpreterWithCustomOps`](https://github.com/tensorflow/tensorflow/blob/547bd9c88b1a86f0543fff3460e2d4d1c8009cb4/tensorflow/lite/python/interpreter.py#L568). We could stabilize it and provide the reference kernels as a kind of custom ops (though it may be confusing as the reference kernels are not really custom ops). 
* Instead we could mimic more closely the C++ API and provide a `tf.lite.InterpreterWithCustomOpResolver` (or add an extra `op_resolver` parameter in `tf.lite.Interpreter`). We could then remove the experimental `tf.lite.InterpreterWithCustomOps` as we could easily register custom ops in the custom op resolver.

I am personally more in favour of a `tf.lite.InterpreterWithCustomOpResolver` as it would not clutter the `tf.lite.Interpreter` while providing a good flexibility and be more general than `tf.lite.InterpreterWithCustomOps`. It would also be quite close to the [`InterpreterBuilder`](https://www.tensorflow.org/lite/api_docs/cc/class/tflite/interpreter-builder) of the C++ API.

**Comment by jdduke** [comment]
There has also been an effort to make it easer to inject a delegate into the Python interpreter, and we could make a pretty straightforward delegate that maps to an OpResolver (in this case the reference OpResolver). Of course, we'd still need to expose and bundle that delegate somehow (or the OpResolver), so there's still a question of whether we'd need an extra API (or if we just implicitly link in the reference ops as a separate delegate .so that gets shipped).

@dmitriykovalev FYI.

**Comment by jdduke** [comment]
Hi @Tessil , we have an internal change prepared for this, but we wanted to check with you first before landing. In our CL, we simply expose the option to use the Reference OpResolver, but we don't provide any finer-grained control (e.g., tweaking the actual registered kernels, or selecting their versions). Is that sufficient for your needs?

**Comment by Tessil** [comment]
Hi, 

Yes it is perfectly sufficient for our needs. We don't need a per-operator fine-tuning of the registered kernels, our main goal is to have a way to run a model using the reference kernels with the Python API.

**Comment by multiverse-tf** [comment]
> Hi,
> 
> Yes it is perfectly sufficient for our needs. We don't need a per-operator fine-tuning of the registered kernels, our main goal is to have a way to run a model using the reference kernels with the Python API.

Great to know about this! Our CL is almost ready, and should be checked in by 3/12.

**Comment by multiverse-tf** [comment]
> > Hi,
> > Yes it is perfectly sufficient for our needs. We don't need a per-operator fine-tuning of the registered kernels, our main goal is to have a way to run a model using the reference kernels with the Python API.
> 
> Great to know about this! Our CL is almost ready, and should be checked in by 3/12.

Hi @Tessil, the support of using reference op resolver in Python has been pushed. Pls refer to
https://github.com/tensorflow/tensorflow/commit/6886228c611131432e550beefce79770d8538972 for details. In short, to use it, when creating the tflite.Interpreter, pass an additional arg as "experimental_op_resolver=OpResolver.BUILTIN_REF".

**Comment by Tessil** [comment]
Hello @multiverse-tf ,

Thank you very much for the commit. 

Small problem I think, though it may be the intended behaviour. 

I can use it with:

```python
import tensorflow as tf
from tensorflow.lite.python.interpreter import OpResolver

interpreter = tf.lite.Interpreter(model, experimental_op_resolver=OpResolver.BUILTIN_REF)
```

but not as I would have expected from the `@_tf_export('lite.experimental.OpResolver')`:

```python
import tensorflow as tf

interpreter = tf.lite.Interpreter(model, experimental_op_resolver=tf.lite.experimental.OpResolver.BUILTIN_REF)
```

I think that in https://github.com/tensorflow/tensorflow/blob/0a150957c4a23e7fd26b2df26e01af661f22a0b2/tensorflow/lite/python/lite.py#L52
a `from tensorflow.lite.python.interpreter import OpResolver  # pylint: disable=unused-import` line should be added, no?

**Comment by multiverse-tf** [comment]
> Hello @multiverse-tf ,
> 
> Thank you very much for the commit.
> 
> Small problem I think, though it may be the intended behaviour.
> 
> I can use it with:
> 
> ```python
> import tensorflow as tf
> from tensorflow.lite.python.interpreter import OpResolver
> 
> interpreter = tf.lite.Interpreter(model, experimental_op_resolver=OpResolver.BUILTIN_REF)
> ```
> 
> but not as I would have expected from the `@_tf_export('lite.experimental.OpResolver')`:
> 
> ```python
> import tensorflow as tf
> 
> interpreter = tf.lite.Interpreter(model, experimental_op_resolver=tf.lite.experimental.OpResolver.BUILTIN_REF)
> ```
> 
> I think that in
> 
> https://github.com/tensorflow/tensorflow/blob/0a150957c4a23e7fd26b2df26e01af661f22a0b2/tensorflow/lite/python/lite.py#L52
> 
> 
> a `from tensorflow.lite.python.interpreter import OpResolver # pylint: disable=unused-import` line should be added, no?
> Otherwise everything seems to work well, thanks.

Thx for the verification! Regarding the small issue you mentioned, just checked in https://github.com/tensorflow/tensorflow/commit/fafba9e833f963bbb8f434f9e0f3f6f557d60d5d to fix it. Note that we've decided to change OpResolver to OpResolverType so that it's more clear semantically. Let me know if there're any other issues.

**Comment by Tessil** [comment]
Thank you, it's working perfectly well now. I'll close the ticket then.
